File: android/ekyc-sdk/src/main/kotlin/com/iapp/ekyc/IappEkycResult.kt
package com.iapp.ekyc

import android.os.Parcelable
import kotlinx.parcelize.Parcelize

/** Flow error (docs/WEBVIEW_BRIDGE.md error codes + engine i18n key). */
@Parcelize
class IappEkycError(
    val code: EkycErrorCode,
    /** HTTP status code when the error came from the API, else null. */
    val statusCode: Int?,
    /** Engine i18n key (resolvable through the engine message tables). */
    val messageKey: String,
    override val message: String,
    /** Retry-After seconds for [EkycErrorCode.RATE_LIMITED]. */
    val retryAfterSeconds: Int? = null,
    /** Fail reason for [EkycErrorCode.LIVENESS_FAILED] (timeout / ...). */
    val reason: String? = null,
) : Exception(message), Parcelable

/**
 * Outcome of a flow. Exactly one subclass per launch: a success variant,
 * [Failed], or [Cancelled].
 */
sealed class IappEkycResult {
    /** Document captured + OCR completed. */
    class DocumentCaptured(
        /** Wire document type, e.g. `"thaiIdFront"`. */
        val documentType: String,
        /** Full OCR response JSON, untouched. */
        val rawJson: String,
        /** Perspective-corrected JPEG (null when `returnSelfieImage=false`). */
        val capturedImage: ByteArray?,
    ) : IappEkycResult()

    /**
     * Active liveness finalized. Only the server-signed [verdictJson]
     * (verify [signature] on YOUR backend) proves liveness — never trust
     * [passed] alone on-device (docs/SECURITY.md).
     */
    class LivenessPassed(
        /** Full finalize response JSON, untouched. */
        val rawJson: String,
        /** The signed verdict object (session_id, selfie_sha256, nonce, ...). */
        val verdictJson: String,
        val passed: Boolean,
        /** hex(HMAC-SHA256(secret, canonicalJSON(verdict))) */
        val signature: String,
        val signatureAlg: String,
        /** The uploaded selfie JPEG (null when `returnSelfieImage=false`). */
        val selfieImage: ByteArray?,
    ) : IappEkycResult()

    /** Frontal selfie captured (no liveness, no API call). */
    class FaceCaptured(val image: ByteArray) : IappEkycResult()

    /** The flow failed — see [error]. */
    class Failed(val error: IappEkycError) : IappEkycResult()

    /** The user cancelled (in-flow Cancel button or system back). */
    object Cancelled : IappEkycResult()
}

/** Simple callback alternative to [IappEkycContract] (see [IappEkyc.start]). */
interface IappEkycCallback {
    /** [IappEkycResult.DocumentCaptured] / [IappEkycResult.LivenessPassed] / [IappEkycResult.FaceCaptured]. */
    fun onResult(result: IappEkycResult)

    fun onError(error: IappEkycError)

    fun onCancelled()
}

File: ios/Sources/IappEkyc/IappEkycResult.swift
import Foundation
import UIKit

/// Result of a `documentCapture` flow: the full OCR response plus the
/// perspective-corrected JPEG the engine uploaded.
@objc(IappEkycDocumentResult)
public final class IappEkycDocumentResult: NSObject {
    /// Wire document type, e.g. `"thaiIdFront"`.
    @objc public let documentType: String
    /// Full parsed OCR response, untouched.
    @objc public let rawJSON: [String: Any]
    /// JPEG bytes of the captured document (nil when `returnSelfieImage=false`).
    @objc public let capturedImageData: Data?
    @objc public var capturedImage: UIImage? {
        capturedImageData.flatMap(UIImage.init(data:))
    }

    init(documentType: String, rawJSON: [String: Any], capturedImageData: Data?) {
        self.documentType = documentType
        self.rawJSON = rawJSON
        self.capturedImageData = capturedImageData
    }
}

/// Result of an `activeLiveness` flow. Only the server-signed `verdictJSON`
/// (verify `signature` on YOUR backend) proves liveness — never trust
/// `passed` alone on-device (docs/SECURITY.md).
@objc(IappEkycLivenessResult)
public final class IappEkycLivenessResult: NSObject {
    /// Full parsed finalize response, untouched.
    @objc public let rawJSON: [String: Any]
    /// The signed verdict object (session_id, selfie_sha256, nonce, ...).
    @objc public let verdictJSON: [String: Any]
    /// Convenience mirror of `verdictJSON["passed"]`.
    @objc public let passed: Bool
    /// hex(HMAC-SHA256(secret, canonicalJSON(verdict)))
    @objc public let signature: String
    @objc public let signatureAlg: String
    /// The selfie JPEG the engine uploaded (nil when `returnSelfieImage=false`).
    @objc public let selfieImageData: Data?
    @objc public var selfieImage: UIImage? {
        selfieImageData.flatMap(UIImage.init(data:))
    }

    init(
        rawJSON: [String: Any],
        verdictJSON: [String: Any],
        passed: Bool,
        signature: String,
        signatureAlg: String,
        selfieImageData: Data?
    ) {
        self.rawJSON = rawJSON
        self.verdictJSON = verdictJSON
        self.passed = passed
        self.signature = signature
        self.signatureAlg = signatureAlg
        self.selfieImageData = selfieImageData
    }
}

/// Result of a `faceCapture` flow (cropped frontal selfie, no liveness).
@objc(IappEkycFaceResult)
public final class IappEkycFaceResult: NSObject {
    @objc public let imageData: Data
    @objc public var image: UIImage? { UIImage(data: imageData) }

    init(imageData: Data) {
        self.imageData = imageData
    }
}

/// Union result: exactly one of `document` / `liveness` / `face` is non-nil,
/// matching `flow`.
@objc(IappEkycResult)
public final class IappEkycResult: NSObject {
    @objc public let flow: IappEkycFlowType
    @objc public let document: IappEkycDocumentResult?
    @objc public let liveness: IappEkycLivenessResult?
    @objc public let face: IappEkycFaceResult?

    init(
        flow: IappEkycFlowType,
        document: IappEkycDocumentResult? = nil,
        liveness: IappEkycLivenessResult? = nil,
        face: IappEkycFaceResult? = nil
    ) {
        self.flow = flow
        self.document = document
        self.liveness = liveness
        self.face = face
    }
}
